Support for XQuery Tumbling and Sliding Window

adamretter opened the pull request:
This implements both `tumbling window` and `sliding window`.

Along the way to make sure it passes the Window tests in the XQTS. We also had to fix a number of other things:

1. Several `fn:deep-equal` bug-fixes
2. Only evaluate arguments to `fn:codepoint-equal` once
3. Complete re-implementation of `fn:sort`
4. Almost a complete rewrite of `group by` expression implementation
5. Fix comparison of `NaN` values for `xs:float` and `xs:double`
6. Fix comparison of `INF` and `-INF` values for `xs:float` and `xs:double`
7. Fix and implement `op:same-key#2` for XQuery Map types
8. Re-implement `fn:deep-equal` in terms of `op:deepCompare`
9. Several Error Code fixes
10. Several XDM Type fixes

By running exist-xqts-runner (from its `compat/eXist-7.0.0-SNAPSHOT` branch):
```bash
exist-xqts-runner-assembly-1.4.0-SNAPSHOT.jar --xqts-version HEAD --output-dir /tmp/xqts-output --test-case fn-codepoint-equal,fn-deep-equal,fn-sort,prod-GroupByClause,prod-WindowClause
```
We see the results:

![Screenshot_select-area_20220827145105](https://user-images.githubusercontent.com/1264057/187275101-8bcfc244-fecb-4fdd-b1e2-1f924c18d5d0.png)

The 7 remaining Window failures are all related to the `count` clause which still needs to be implemented first in a separate PR, see: https://github.com/eXist-db/exist/pull/4530.

----
This open source contribution to the eXist-db project was commissioned and in part sponsored by the Office of the Historian, U.S. Department of State, https://history.state.gov/, the remaining sponsorship and implementation came from [Evolved Binary](https://ww.evolvedbinary.com)

joewiz commented:
Very exciting! This works with all of the code from _XQuery for Humanists_ in the sections on Windowing except for one, which apparently isn't tested in XQTS - and which I'll submit as a new test there. This query is:

```xquery
let $sentence := "This is a a test of of windowing"
let $words := fn:tokenize($sentence, " ")
for tumbling window $window in $words
    start $ws at $wsn previous $wp next $wn 
        when $wp eq $ws
return
    <window ws="{$ws}" wsn="{$wsn}" wp="{$wp}" wn="{$wn}">{ $window }</window>
```

The results from this branch:

```xml
<window ws="of" wsn="7" wp="of" wn="windowing">a test of</window>
<window ws="windowing" wsn="8" wp="of" wn="">of windowing</window>
```

The expected results, as returned by BaseX and Saxon:

```xml
<window ws="a" wsn="4" wp="a" wn="test">a test of</window>
<window ws="of" wsn="7" wp="of" wn="windowing">of windowing</window>
```

(The window is selected correctly, but the variables bound by the `start`, `previous`, `next`, and `at` clauses are wrong.)

Also, there seems to be one parser issue. The query:

```xquery
<x/>/@when
```

... returns the error:

> error found while executing expression: org.exist.xquery.XPathException: err:XPST0003 unexpected token: when [at line 1, column 7]

From exist.log:

```
2022-08-30 08:53:17,523 [qtp1304614894-67] ERROR (XQueryServlet.java [process]:549) - Cannot compile xquery: err:XPST0003 org.exist.xquery.XPathException: err:XPST0003 unexpected token: when [at line 1, column 7]
 [at line 7, column 1] 
org.exist.EXistException: Cannot compile xquery: err:XPST0003 org.exist.xquery.XPathException: err:XPST0003 unexpected token: when [at line 1, column 7]
 [at line 7, column 1]
...
Caused by: org.exist.xquery.StaticXQueryException: err:XPST0003 org.exist.xquery.XPathException: err:XPST0003 unexpected token: when [at line 1, column 7]
 [at line 7, column 1]
	at org.exist.xquery.XQuery.compile(XQuery.java:235) ~[exist-core-6.1.0-SNAPSHOT.jar:6.1.0-SNAPSHOT]
	at org.exist.xquery.XQuery.compile(XQuery.java:180) ~[exist-core-6.1.0-SNAPSHOT.jar:6.1.0-SNAPSHOT]
	at org.exist.xquery.XQuery.compile(XQuery.java:138) ~[exist-core-6.1.0-SNAPSHOT.jar:6.1.0-SNAPSHOT]
	at org.exist.http.servlets.XQueryServlet.process(XQueryServlet.java:439) ~[exist-core-6.1.0-SNAPSHOT.jar:6.1.0-SNAPSHOT]
	... 66 more
```

joewiz commented:
@adamretter I can confirm that with the last commit the `<x/>/@when` issue is no longer present. Thank you!

dizzzz commented on exist-core/src/main/java/org/exist/xquery/AbstractFLWORClause.java:
```diff
@@ -41,14 +41,10 @@ public AbstractFLWORClause(XQueryContext context) {
-        }
+    public LocalVariable createVariable(final QName name) throws XPathException {
+        final LocalVariable var = new LocalVariable(name);
+        firstVar = var;
```
return var immediately?

adamretter replied:
I think `firstVar` has to be set, as the variables kind of form a linked list of sorts in the state of the AbstractFLWORClause

duncdrum reviewed (commented):
Nice change. This needs a rebase to resolve conflicts. It’s there content work still missing to no longer mark it as draft?

adamretter commented:
@joewiz This now passes your `twt:jw-duplicate-words#0` test.

adamretter commented:
@dizzzz @reinhapa I think this one is now ready to land. Let me know your thoughts....

joewiz commented:
@adamretter It appears there is a failing test. From [ubuntu-latest Test](https://github.com/eXist-db/exist/actions/runs/5193725174/jobs/9364630486?pr=4529#step:6:1698) - see also [macOS-latest Test](https://github.com/eXist-db/exist/actions/runs/5193725174/jobs/9364630694?pr=4529#step:6:708):

```
Error:  Tests run: 833, Failures: 1, Errors: 0, Skipped: 4, Time elapsed: 50.683 s <<< FAILURE! - in xquery.xquery3.XQuery3Tests
Error:  xqts.org.exist-db.xquery.test.groupby.atomize-group-vars  Time elapsed: 0.103 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<...g-time&gt;&lt;person[&gt;Anton&lt;/person&gt;&lt;other&gt;&lt;person&gt;Anton&lt;/person&gt;&lt;/other&gt;&lt;/working-time&gt;&lt;working-time&gt;&lt;person&gt;Barbara&lt;/person&gt;&lt;other&gt;&lt;person&gt;Barbara&lt;/person&gt;&lt;/other&gt;&lt;/working-time&gt;&lt;working-time&gt;&lt;person&gt;Clara&lt;/person&gt;&lt;other&gt;&lt;person&gt;Clara&lt;/person&gt;&lt;/other&gt;&lt;/working-time&gt;&lt;working-time&gt;&lt;person/&gt;&lt;other/&gt;&lt;/working-time&gt;&lt;/result&gt; (assertEquals failed.)]> but was:<...g-time&gt;&lt;person[/&gt;&lt;other/&gt;&lt;/working-time&gt;&lt;working-time&gt;&lt;person&gt;Anton&lt;/person&gt;&lt;other&gt;&lt;person&gt;Anton&lt;/person&gt;&lt;/other&gt;&lt;/working-time&gt;&lt;working-time&gt;&lt;person&gt;Barbara&lt;/person&gt;&lt;other&gt;&lt;person&gt;Barbara&lt;/person&gt;&lt;/other&gt;&lt;/working-time&gt;&lt;working-time&gt;&lt;person&gt;Clara&lt;/person&gt;&lt;other&gt;&lt;person&gt;Clara&lt;/person&gt;&lt;/other&gt;&lt;/working-time&gt;&lt;/result&gt;]>
```

Examining the failing test, https://github.com/eXist-db/exist/blob/e72d7368dd0ae0fb7656ee3985893b3ac3ae846d/exist-core/src/test/xquery/xquery3/groupby.xql#L633-L720, the test is written to expect the following output:

```xml
<result>
    <working-time>
        <person>Anton</person>
        <other>
            <person>Anton</person>
        </other>
    </working-time>
    <working-time>
        <person>Barbara</person>
        <other>
            <person>Barbara</person>
        </other>
    </working-time>
    <working-time>
        <person>Clara</person>
        <other>
            <person>Clara</person>
        </other>
    </working-time>
    <working-time>
        <person/>
        <other/>
    </working-time>
</result>
```

However, it actually returns the following:

```xml
<result>
    <working-time>
        <person/>
        <other/>
    </working-time>
    <working-time>
        <person>Anton</person>
        <other>
            <person>Anton</person>
        </other>
    </working-time>
    <working-time>
        <person>Barbara</person>
        <other>
            <person>Barbara</person>
        </other>
    </working-time>
    <working-time>
        <person>Clara</person>
        <other>
            <person>Clara</person>
        </other>
    </working-time>
</result>
```

(The difference is in the position of the `<working-time>` element with the empty `<person>` and `<other>` child elements.)

The actual results also match the results of BaseX and Saxon. 

Thus, I presume the test should be updated to match the actual results:

```xquery
%test:assertEquals("<result><working-time><person/><other/></working-time><working-time><person>Anton</person><other><person>Anton</person></other></working-time><working-time><person>Barbara</person><other><person>Barbara</person></other></working-time><working-time><person>Clara</person><other><person>Clara</person></other></working-time></result>")
```

adamretter commented:
@joewiz Thanks I fixed that bad test.
This is not the first one of these I have found, I do wonder how people ended up adding tests in the past that don't comply with the XQuery specs!?!

joewiz commented:
@adamretter Thanks! I think that test was introduced in this PR - [before](https://github.com/eXist-db/exist/blob/develop/exist-core/src/test/xquery/xquery3/groupby.xql#L631) vs. [after](https://github.com/eXist-db/exist/blame/a0bcbd5c2d81d997555ebc93ca3c1caa717a0819/exist-core/src/test/xquery/xquery3/groupby.xql#L633-L720).

adamretter commented:
@joewiz Ah yes, you are correct!

It has been an insane amount of work to fix so many broken things in eXist-db just so I could get to the point of being able to implement and test the Window expressions. I have been working on this for almost 1 year (since June 2022), and so there are a lot of things that I have fixed and forgotten about along the way ;-)

reinhapa commented on exist-core/src/main/java/org/exist/xquery/WindowExpr.java:
```diff
@@ -0,0 +1,684 @@
+    @Override
+    public Set<QName> getTupleStreamVariables() {
+        final Set<QName> vars = new HashSet<>();
+        if (varName != null) {
+            vars.add(varName);
+        }
+
+        if (windowStartCondition.getPreviousItem() != null) {
+            vars.add(windowStartCondition.getPreviousItem());
+        }
+        if (windowStartCondition.getCurrentItem() != null) {
+            vars.add(windowStartCondition.getCurrentItem());
+        }
+        if (windowStartCondition.getNextItem() != null) {
+            vars.add(windowStartCondition.getNextItem());
+        }
+
+        if (windowEndCondition != null) {
+            if (windowEndCondition.getPreviousItem() != null) {
+                vars.add(windowEndCondition.getPreviousItem());
+            }
+            if (windowEndCondition.getCurrentItem() != null) {
+                vars.add(windowEndCondition.getCurrentItem());
+            }
+            if (windowEndCondition.getNextItem() != null) {
+                vars.add(windowEndCondition.getNextItem());
+            }
+        }
+
+        final QName var = getVariable();
+        if (var != null) {
+            vars.add(var);
+        }
+
+        final LocalVariable startVar = getStartVariable();
+        if (startVar != null) {
+            vars.add(startVar.getQName());
+        }
+
```
Maby extracting the check for null and add into a simple function would make this method more compact with something like this:
```
    @Override
    public Set<QName> getTupleStreamVariables() {
        final Set<QName> vars = new HashSet<>();
        addIfNotNull(vars, varName);
        addIfNotNull(vars, windowStartCondition.getPreviousItem());
        addIfNotNull(vars, windowStartCondition.getCurrentItem());
        addIfNotNull(vars, windowStartCondition.getNextItem());

        if (windowEndCondition != null) {
           addIfNotNull(vars, windowEndCondition.getPreviousItem());
           addIfNotNull(vars, windowEndCondition.getCurrentItem());
           addIfNotNull(vars, windowEndCondition.getNextItem());
        }

        addIfNotNull(vars, getVariable());

        final LocalVariable startVar = getStartVariable();
        if (startVar != null) {
            vars.add(startVar.getQName());
        }

        return vars;
    }

    private void addIfNotNull(Set<QName> vars, String value) {
      if (value != null) {
        vars.add(value);
      }
    }
```

adamretter replied:
Done.

reinhapa commented on exist-core/src/main/java/org/exist/xquery/functions/fn/CollatingFunction.java:
```diff
@@ -22,10 +22,7 @@
-import org.exist.xquery.FunctionSignature;
-import org.exist.xquery.XPathException;
-import org.exist.xquery.XQueryContext;
+import org.exist.xquery.*;
```
Do we not prefer single imports?

dizzzz replied:
pooh, different context I see different 'rules'

dizzzz replied:
this is std IntelliJ

reinhapa commented on exist-core/src/main/java/org/exist/xquery/functions/fn/FunDeepEqual.java:
```diff
@@ -130,218 +129,344 @@ public Sequence eval(Sequence contextSequence, Item contextItem)
+                    return safeCompare(nva.getStringValue(), nvb.getStringValue(), collator);
+
+                default:
+                    throw new RuntimeException("unexpected item type " + Type.getTypeName(item1.getType()));
```
Would an `IllegalArgumentException` not more concise here?

reinhapa commented on exist-core/src/main/java/org/exist/xquery/functions/fn/FunDeepEqual.java:
```diff
@@ -351,16 +476,34 @@ private static int getAttrCount(NamedNodeMap nnm) {
-        return Objects.equals(a, b);
-    }
+    private static int safeCompare(@Nullable final String a, @Nullable final String b, @Nullable final Collator collator) {
+        if (a == b) {
```
Should not be `Objects.equals()` used here?

reinhapa commented on exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java:
```diff
@@ -146,19 +146,33 @@ public boolean isPositive() {
+            comparison = () -> other.isNegativeInfinity() ? Constants.EQUAL : Constants.INFERIOR;
+        } else if (other.isNegativeInfinity()) {
+            comparison = () -> Constants.SUPERIOR;
+        } else if (other instanceof IntegerValue) {
+            comparison = () -> BigDecimal.valueOf(value).compareTo(new BigDecimal(((IntegerValue) other).value));
+        } else if (other instanceof DecimalValue) {
+            comparison = () -> BigDecimal.valueOf(value).compareTo(((DecimalValue) other).value);
+        } else if (other instanceof DoubleValue) {
+            comparison = () -> Double.compare(value, ((DoubleValue) other).value);
+        } else if (other instanceof FloatValue) {
+            comparison = () -> Double.compare(value, ((FloatValue) other).value);
```
With JDK 17 the simpler pattern cast could be used here:
```
        } else if (other instanceof IntegerValue integerValue) {
            comparison = () -> BigDecimal.valueOf(value).compareTo(new BigDecimal(integerValue.value));
        } else if (other instanceof DecimalValue decimalValue) {
            comparison = () -> BigDecimal.valueOf(value).compareTo(decimalValue.value);
        } else if (other instanceof DoubleValue doubleValue) {
            comparison = () -> Double.compare(value, doubleValue.value);
        } else if (other instanceof FloatValue floatValue) {
            comparison = () -> Double.compare(value, floatValue.value);
``

reinhapa commented on exist-core/src/main/java/org/exist/xquery/value/OrderedValueSequence.java:
```diff
@@ -400,20 +537,27 @@ public int compareTo(final Entry other) {
+                return item.toString();
+            } else {
+                final StringBuilder builder = new StringBuilder();
+                builder.append(item);
+                builder.append(" [");
+                for (int i = 0; i < values.size(); i++) {
+                    if (i > 0) {
+                        builder.append(", ");
+                    }
+                    builder.append(values.get(i).toString());
                 }
-                builder.append(values[i].toString());
+                builder.append("]");
+                return builder.toString();
```
Using a `StringJoiner` may make this even more readable...

reinhapa commented on exist-core/src/test/java/org/exist/xquery/WindowClauseTest.java:
```diff
@@ -0,0 +1,474 @@
+        }
+
+        assertTrue(expr.getFirst() instanceof WindowExpr, "Expression should be of type WindowExpr");
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getCurrentItem());
+        assertEquals(new QName("s"), ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPosVar());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPreviousItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getNextItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowEndCondition().getCurrentItem());
+        assertEquals(new QName("e"), ((WindowExpr) expr.getFirst()).getWindowEndCondition().getPosVar());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowEndCondition().getPreviousItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowEndCondition().getNextItem());
+        assertEquals(true, ((WindowExpr) expr.getFirst()).getWindowEndCondition().isOnly());
```
Any special reason not using `assertTrue` rsp. `assertNull` for the boolean and null cases?

reinhapa commented on exist-core/src/test/java/org/exist/xquery/WindowClauseTest.java:
```diff
@@ -0,0 +1,474 @@
+        }
+
+        assertTrue(expr.getFirst() instanceof WindowExpr, "Expression should be of type WindowExpr");
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getCurrentItem());
+        assertEquals(new QName("s"), ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPosVar());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPreviousItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getNextItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowEndCondition());
```
Any special reason not using `assertNull` for the null cases?

reinhapa commented on exist-core/src/test/java/org/exist/xquery/WindowClauseTest.java:
```diff
@@ -0,0 +1,474 @@
+
+        assertTrue(expr.getFirst() instanceof WindowExpr, "Expression should be of type WindowExpr");
+        assertEquals("first", ((WindowExpr) expr.getFirst()).getWindowStartCondition().getCurrentItem().getStringValue());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPosVar());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPreviousItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getNextItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowEndCondition());
```
Any special reason not using `assertNull` for the null cases?

reinhapa commented on exist-core/src/test/java/org/exist/xquery/WindowClauseTest.java:
```diff
@@ -0,0 +1,474 @@
+
+        assertTrue(expr.getFirst() instanceof WindowExpr, "Expression should be of type WindowExpr");
+        assertEquals(WindowExpr.WindowType.SLIDING_WINDOW, ((WindowExpr) expr.getFirst()).getWindowType());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getCurrentItem());
+        assertEquals(new QName("s"), ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPosVar());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPreviousItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getNextItem());
+        assertEquals(new QName("e"), ((WindowExpr) expr.getFirst()).getWindowEndCondition().getPosVar());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowEndCondition().getPreviousItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowEndCondition().getNextItem());
+        assertEquals(true, ((WindowExpr) expr.getFirst()).getWindowEndCondition().isOnly());
```
Any special reason not using `assertTrue` rsp. `assertNull` for the boolean and null cases?

reinhapa commented on exist-core/src/test/java/org/exist/xquery/WindowClauseTest.java:
```diff
@@ -0,0 +1,474 @@
+
+        assertTrue(expr.getFirst() instanceof WindowExpr, "Expression should be of type WindowExpr");
+        assertEquals(WindowExpr.WindowType.SLIDING_WINDOW, ((WindowExpr) expr.getFirst()).getWindowType());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getCurrentItem());
+        assertEquals(new QName("s"), ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPosVar());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPreviousItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getNextItem());
+        assertEquals(new QName("e"), ((WindowExpr) expr.getFirst()).getWindowEndCondition().getPosVar());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowEndCondition().getPreviousItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowEndCondition().getNextItem());
+        assertEquals(false, ((WindowExpr) expr.getFirst()).getWindowEndCondition().isOnly());
```
Any special reason not using `assertFalse` rsp. `assertNull` for the boolean and null cases?

reinhapa commented on exist-core/src/test/java/org/exist/xquery/WindowClauseTest.java:
```diff
@@ -0,0 +1,474 @@
+        assertTrue(expr.getFirst() instanceof WindowExpr, "Expression should be of type WindowExpr");
+        assertEquals(WindowExpr.WindowType.TUMBLING_WINDOW, ((WindowExpr) expr.getFirst()).getWindowType());
+        assertEquals(new QName("first"), ((WindowExpr) expr.getFirst()).getWindowStartCondition().getCurrentItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPosVar());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowStartCondition().getPreviousItem());
+        assertEquals(new QName("second"), ((WindowExpr) expr.getFirst()).getWindowStartCondition().getNextItem());
+        assertEquals(new QName("last"), ((WindowExpr) expr.getFirst()).getWindowEndCondition().getCurrentItem());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowEndCondition().getPosVar());
+        assertEquals(null, ((WindowExpr) expr.getFirst()).getWindowEndCondition().getPreviousItem());
+        assertEquals(new QName("beyond"), ((WindowExpr) expr.getFirst()).getWindowEndCondition().getNextItem());
+        assertEquals(false, ((WindowExpr) expr.getFirst()).getWindowEndCondition().isOnly());
```
Any special reason not using `assertFalse` rsp. `assertNull` for the boolean and null cases?

reinhapa commented on exist-core/src/test/xquery/xquery3/serialize.xql:
```diff
@@ -254,7 +254,7 @@ function ser:serialize-empty-sequence() {
 declare
     %test:args(1234)
-    %test:assertEquals('1234')
+    %test:assertEquals(1234)
```
Seems to be unrelated to tumbling window

adamretter replied:
It is simply reseting something that has been incorrectly changed since I forked the code to work on Window Expressions.

joewiz replied:
@adamretter Patrick raises a valid concern here that the changes to serialize.xql in https://github.com/eXist-db/exist/commit/ccfb1d5a38fc4a882968f68c7165744ac1f088eb have no relation the present PR. Please remove this commit from the PR. If the change is needed, it should be addressed in a separate PR.

adamretter replied:
@joewiz Those changes were previously part of a larger set of serialization fixes that were needed for this PR to correctly pass the XQTS. Most of those serialization fixes I extracted and sent in a separate PR that was previously merged. These appear here as an artifact of rebasing on those previously merged changes.

For the sake of getting this merged, I have now removed that commit as you and Patrick requested.

reinhapa commented on exist-core/src/test/xquery/xquery3/serialize.xql:
```diff
@@ -265,7 +265,7 @@ function ser:adaptive-simple-atomic($atomic as xs:anyAtomicType) {
 declare
     %test:args(1234)
-    %test:assertEquals('1234')
+    %test:assertEquals(1234)
```
Seems to be unrelated to tumbling window

adamretter replied:
It is simply reseting something that has been incorrectly changed since I forked the code to work on Window Expressions.

reinhapa commented on exist-core/src/main/java/org/exist/xquery/ForExpr.java:
```diff
@@ -26,14 +26,19 @@
 import org.exist.xquery.value.*;
 
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
```
Unused imports should be removed

adamretter replied:
Done.

reinhapa commented on exist-core/src/main/java/org/exist/xquery/ForExpr.java:
```diff
@@ -26,14 +26,19 @@
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
```
Unused imports should be removed

adamretter replied:
Done.

reinhapa commented on exist-core/src/main/java/org/exist/xquery/GroupByClause.java:
```diff
@@ -185,11 +189,31 @@ public Sequence postEval(final Sequence seq) throws XPathException {
+
+                    if (spec.getGroupExpression() instanceof VariableReference) {
+                        final VariableReference groupExpressionVariableRef = ((VariableReference) spec.getGroupExpression());
+                        if (groupExpressionVariableRef.getName().equals(keyVarName)) {
+                            if (!preGroupingTupleVariables.contains(keyVarName)) {
```
Any reason not joining this expressions?

adamretter replied:
Done.

adamretter commented on exist-core/src/main/java/org/exist/xquery/WindowExpr.java:
```diff
@@ -0,0 +1,684 @@
+                } else {
+                    endWhen = false;
+                }
+                if (endWhen || (windowType == WindowType.SLIDING_WINDOW && (!windowEndCondition.isOnly()) && i == inCount - 1 && i > windowStartIdx)) {
+
+                    if (window != null) {
```
The two if statements are checking for quite different conditions and so when working through how this code works, it it reduces the cognitive load on the developer to keep them separate.

adamretter commented on exist-core/src/main/java/org/exist/xquery/WindowCondition.java:
```diff
@@ -0,0 +1,128 @@
+        if (result.length() != 0) {
+            result.append(' ');
+        }
+        result.append("when ").append(whenExpression);
```
`StringJoiner` won't work here, as each component has a different prefix. To address the NPath complexity warning from Codacy, I have instead refactored this to abstract out the common pattern.

adamretter commented on exist-core/src/main/java/org/exist/xquery/WindowExpr.java:
```diff
@@ -0,0 +1,684 @@
+
+        // pre-eval the return expr if it is a FLWORClause and we are the outer for-loop
+        if (isOuterFor) {
+            if (returnExpr instanceof FLWORClause) {
```
Done.

adamretter commented:
@reinhapa Thanks for the review :-) I think I have addressed your concerns now, and so I have requested a further review...

reinhapa commented on exist-core/src/main/java/org/exist/xquery/GroupByClause.java:
```diff
@@ -185,11 +189,29 @@ public Sequence postEval(final Sequence seq) throws XPathException {
-                    final LocalVariable groupKeyVar = new LocalVariable(spec.getKeyVarName());
+                    final QName keyVarName = spec.getKeyVarName();
+
+                    if (spec.getGroupExpression() instanceof VariableReference groupExpressionVariableRef) {
+                        if (groupExpressionVariableRef.getName().equals(keyVarName)
```
any reason to not combine this statements?

adamretter replied:
Purely for the purposes of making the code easier to read

dizzzz commented on exist-core/src/main/antlr/org/exist/xquery/parser/XQuery.g:
```diff
@@ -103,6 +103,10 @@ options {
 	}
 
+	public Exception getLastException() {
+        return (Exception) exceptions.get(exceptions.size() - 1);
```
size can't return 0?

adamretter replied:
It is actually only every called if `XQueryParser#foundErrors()` returns true first, however just to make the intention clear I added and additional check there to avoid a potential `IndexOutOfBoundsException`

adamretter commented:
@dizzzz @reinhapa I believe this is now ready to merge please

reinhapa commented:
@adamretter could you address some more of those Codacy/SonarCloud issues? There seem a lot of simple ones that can be solved easy...

adamretter commented:
@reinhapa each of the 5 bugs listed by SonarCloud are false positives. I fixed a few small code smell issues highlighted by SonarCloud. The remaining suggestions from SonarCloud and Codacy otherwise IMHO make the code worse!

reinhapa commented:
@adamretter there is a compile error on the test source:
```
/D:/a/exist/exist/exist-core/src/main/java/org/exist/xquery/BasicExpressionVisitor.java:[34,8] org.exist.xquery.BasicExpressionVisitor is not abstract and does not override abstract method visitWindowExpression(org.exist.xquery.WindowExpr) in org.exist.xquery.ExpressionVisitor
```

dizzzz reviewed (changes requested):
compilation issue

adamretter commented:
@dizzzz @reinhapa I have fixed the compilation issue (caused by a previous rebase), and again now rebased this. I believe this is now ready to merge please. I would really like to see this merged ASAPP as previously having it left languishing led to a lot of additional work for me to rebase it multiple times!

adamretter commented on exist-core/src/main/java/org/exist/xquery/Context.java:
```diff
@@ -43,21 +43,11 @@
 import org.exist.xmldb.XmldbURI;
-import org.exist.xquery.AnalyzeContextInfo;
-import org.exist.xquery.Expression;
+import org.exist.xquery.*;
```
@dizzzz Thanks, I just appended a commit to remove those.

reinhapa commented on exist-core/src/main/java/org/exist/xquery/ValueComparison.java:
```diff
@@ -130,63 +135,108 @@ protected Sequence nodeSetCompare(NodeSet nodes, Sequence contextSequence) throw
+	 * @return the result of the comparison
+	 * @throws XPathException in case of dynamic error
+	 */
+	public static int compareAtomic(@Nullable final Collator collator, AtomicValue lv, AtomicValue rv) throws XPathException {
```
Some arguments are not final here too..

adamretter replied:
They are mutated below @reinhapa

reinhapa commented on exist-core/src/main/java/org/exist/xquery/ValueComparison.java:
```diff
@@ -130,63 +135,108 @@ protected Sequence nodeSetCompare(NodeSet nodes, Sequence contextSequence) throw
+	 * @return the result of the comparison
+	 * @throws XPathException in case of dynamic error
+	 */
+	private static Tuple2<AtomicValue, AtomicValue> makeComparable(AtomicValue lv, AtomicValue rv, final StringTruncationOperator truncation) throws XPathException {
```
.. again missing final's on arguments

adamretter replied:
They are mutated below @reinhapa

reinhapa commented on exist-core/src/main/java/org/exist/xquery/functions/fn/FunDeepEqual.java:
```diff
@@ -130,218 +129,353 @@ public Sequence eval(Sequence contextSequence, Item contextItem)
     }
 
-    private static boolean compareContents(Node a, Node b) {
+    private static int compareContents(Node a, Node b, @Nullable final Collator collator) {
```
missing final's on arguments

adamretter replied:
They are mutated below @reinhapa

reinhapa commented on exist-core/src/main/java/org/exist/xquery/functions/fn/FunDeepEqual.java:
```diff
@@ -130,218 +129,353 @@ public Sequence eval(Sequence contextSequence, Item contextItem)
+        }
     }
 
     private static Node findNextTextOrElementNode(Node n) {
```
Argument not final

adamretter replied:
It is mutated below @reinhapa

reinhapa commented on exist-core/src/main/java/org/exist/xquery/value/BooleanValue.java:
```diff
@@ -114,14 +114,17 @@ public boolean compareTo(Collator collator, Comparison operator, AtomicValue oth
     }
 
     public int compareTo(Collator collator, AtomicValue other) throws XPathException {
```
Some more missing `final` tags

adamretter replied:
Hmm... this isn't related to my changes. I will add a commit to mark it as a final as you have requested

reinhapa reviewed (changes requested):
There are a couple places where not all record arguments where final...

adamretter commented on exist-core/src/main/java/org/exist/xquery/ValueComparison.java:
```diff
@@ -130,63 +135,108 @@ protected Sequence nodeSetCompare(NodeSet nodes, Sequence contextSequence) throw
-		if (rtype == Type.UNTYPED_ATOMIC) {
-			//If the atomized operand is of type xs:untypedAtomic, it is cast to xs:string.
-			rv = rv.convertTo(Type.STRING);
+	public static boolean compareAtomic(@Nullable final Collator collator, AtomicValue lv, AtomicValue rv, final StringTruncationOperator truncation, final Comparison relation) throws XPathException {
```
No. `lv` and `rv` are mutated below @reinhapa

reinhapa replied:
The reasoning for having all arguments final for me is that we may want to reset to the method frame and having changed arguments at that point would be bad.

adamretter replied:
I always use `final` unless there is a good reason not to. Here I want the most performance that I can get, so I avoid allocating another variable on the stack.

adamretter commented on exist-core/src/main/java/org/exist/xquery/XQueryContext.java:
```diff
@@ -1090,11 +1090,30 @@ public String getDefaultCollation() {
 
     @Override
-    public Collator getCollator(final String uri) throws XPathException {
+    public Collator getCollator(String uri) throws XPathException {
```
No. It is mutated below @reinhapa

adamretter commented on exist-core/src/main/java/org/exist/xquery/BasicExpressionVisitor.java:
```diff
@@ -176,7 +176,7 @@ public void visitLetExpression(final LetExpr letExpr) {
 
     @Override
-    public void visitOrderByClause(final OrderByClause orderBy) {
+    public void visitOrderByClause(OrderByClause orderBy) {
```
Yes that one was an accident. I have now fixed it.

adamretter commented on exist-core/src/main/java/org/exist/xquery/GroupByClause.java:
```diff
@@ -78,7 +74,7 @@ public ClauseType getType() {
 
     @Override
     public Sequence preEval(Sequence seq) throws XPathException {
```
Hmm... this isn't related to my changes. I will add a commit to mark it as a final as you have requested... but where does the 'tidy up' of existing code start and begin?

adamretter commented on exist-core/src/main/java/org/exist/xquery/GroupByClause.java:
```diff
@@ -185,11 +189,29 @@ public Sequence postEval(final Sequence seq) throws XPathException {
     public void analyze(AnalyzeContextInfo contextInfo) throws XPathException {
```
Hmm... this isn't related to my changes. I will add a commit to mark it as a final as you have requested.

adamretter commented:
@reinhapa I hope I have addressed your requests to your satisfaction now. Let me know if there is anything else you need :-)

reinhapa reviewed (approved):
Would be nice to have all method arguments final in the future...
